### PR TITLE
AO3-6647 Upgrade Ruby to 3.1.4

### PR DESCRIPTION
**Pull request by zz9pzza**
# Pull Request Checklist

* [x] Have you read ["How to write the perfect pull request"](https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/)?
* [x] Have you read the [contributing guidelines](https://github.com/otwcode/otwarchive/blob/master/CONTRIBUTING.md)?
* [x] Have you added [tests for any changed functionality](https://github.com/otwcode/otwarchive/wiki/Automated-Testing)?
* [x] Have you added the [Jira](https://otwarchive.atlassian.net) issue number
  as the *first* thing in your pull request title (e.g. `AO3-1234 Fix thing`)
* [x] Do you have fewer than 5 pull requests already open? If not, please wait
  until they are reviewed and merged before creating new pull requests.

## Issue

https://otwarchive.atlassian.net/browse/AO3-6647

## Purpose

Change the version of ruby that the Archive uses.

## Testing Instructions

I would do a smoke test

## References

Are there other relevant issues/pull requests/mailing list discussions?

## Credit

He/Him

### CODE REVIEW

**Comment by zz9pzza** [comment]
Need to check 
`WARNING: The monkeypatch /otwa/config/initializers/monkeypatches/mail_disable_starttls.rb was written for version 2.7.1 of the mail gem, but you are running 2.8.1. Please update or remove the monkeypatch.`

**Comment by Bilka2** [comment]
It may be good to update the [parser gem](https://github.com/whitequark/parser) to at least [v3.2.2.0](https://github.com/whitequark/parser/blob/master/CHANGELOG.md#v3220-2023-04-01) to support ruby 3.1.4.

**Comment by Bilka2** [comment]
The upgraded parser gem is missing from the commit. Rest looks good.

File: Gemfile.lock [hunk 1]
```diff
@@ -603,7 +616,7 @@ GEM
     zeitwerk (2.6.8)
 
 PLATFORMS
-  ruby
+  x86_64-linux
```
Comment (Bilka2) [inline Gemfile.lock:619]: What happened here and in the `nokogiri (1.16.2-x86_64-linux)` line? The change was introduced with the merge from upstream. But nokogiri should not be platform dependent.

Comment (Bilka2) [reply Gemfile.lock:619]: Maybe similar to https://github.com/rubygems/rubygems/issues/7304#issuecomment-1863317056 - removal of `mini_portile2` dependency may have caused the install after the merge to prefer to use the non-generic nokogiri.

**Review by Bilka2 — APPROVED** [review]

**Pull request merged**